jung-algorithms/src/main/java/edu/uci/ics/jung/layout/model/LoadingCacheLayoutModel.java

```diff
@@ -17,47 +17,55 @@
 
-  public LoadingCacheLayoutModel(LoadingCacheLayoutModel<N, P> other) {
-    super(other.graph, other.pointModel, other.width, other.height, other.depth);
-  }
+  public static class Builder<N, P> extends AbstractLayoutModel.Builder<N, P> {
+    protected LoadingCache<N, P> locations =
+        CacheBuilder.newBuilder().build(CacheLoader.from(() -> pointModel.newPoint(0, 0, 0)));
 
-  public LoadingCacheLayoutModel(LoadingCacheLayoutModel<N, P> other, int width, int height) {
-    this(other, width, height, 0);
-  }
+    public LoadingCacheLayoutModel.Builder<N, P> setGraph(Graph<N> graph) {
+      super.setGraph(graph);
+      return this;
+    }
 
-  public LoadingCacheLayoutModel(
-      LoadingCacheLayoutModel<N, P> other, int width, int height, int depth) {
-    super(other.graph, other.pointModel, width, height, depth);
-  }
+    public LoadingCacheLayoutModel.Builder<N, P> setPointModel(PointModel<P> pointModel) {
+      super.setPointModel(pointModel);
+      return this;
+    }
 
-  public LoadingCacheLayoutModel(
-      Graph<N> graph, PointModel<P> pointModel, int width, int height, Function<N, P> initializer) {
-    this(graph, pointModel, width, height, 0, initializer);
-  }
+    public LoadingCacheLayoutModel.Builder<N, P> setSize(int width, int height, int depth) {
+      super.setSize(width, height, depth);
+      return this;
+    }
 
-  public LoadingCacheLayoutModel(
-      Graph<N> graph,
-      PointModel<P> pointModel,
-      int width,
-      int height,
-      int depth,
-      Function<N, P> initializer) {
-    super(graph, pointModel, width, height, depth);
-    Function<N, P> chain =
-        initializer.andThen(
-            p -> pointModel.newPoint(pointModel.getX(p), pointModel.getY(p), pointModel.getZ(p)));
-    this.locations = CacheBuilder.newBuilder().build(CacheLoader.from(chain::apply));
+    public LoadingCacheLayoutModel.Builder<N, P> setSize(int width, int height) {
+      super.setSize(width, height, 0);
+      return this;
+    }
+
+    public LoadingCacheLayoutModel.Builder<N, P> setInitializer(Function<N, P> initializer) {
+
+      Function<N, P> chain =
+          initializer.andThen(
+              p -> pointModel.newPoint(pointModel.getX(p), pointModel.getY(p), pointModel.getZ(p)));
+      this.locations = CacheBuilder.newBuilder().build(CacheLoader.from(chain::apply));
+      return this;
+    }
+
+    public LoadingCacheLayoutModel<N, P> build() {
+      return new LoadingCacheLayoutModel<>(this);
+    }
+
+    public LoadingCacheLayoutModel<N, P> from(LoadingCacheLayoutModel<N, P> other) {
+      return new LoadingCacheLayoutModel<N, P>(other);
+    }
   }
 
-  public LoadingCacheLayoutModel(Graph<N> graph, PointModel<P> pointModel, int width, int height) {
-    this(graph, pointModel, width, height, 0);
+  protected LoadingCacheLayoutModel(LoadingCacheLayoutModel.Builder<N, P> builder) {
+    super(builder.graph, builder.pointModel, builder.width, builder.height, builder.depth);
```

Comment:
Any builder parameters that don't (and can't) have meaningful defaults need to be checked for null (or analogous checks for primitive types), e.g.:

    this.graph = Preconditions.checkNotNull(builder.graph);

(etc.)

Reply:
i agree. I have been thinking about where to put them (in the builder, in the thing it builds, etc) I was looking for examples for the style.